src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -38,43 +38,38 @@ internal static partial class ExtensionMethodImportCompletionHelper
             bool forceIndexCreation,
             CancellationToken cancellationToken)
         {
-            var ticks = Environment.TickCount;
-            var project = document.Project;
-
-            // This service is only defined for C# and VB, but we'll be a bit paranoid.
-            var client = RemoteSupportedLanguages.IsSupported(project.Language)
-                ? await project.Solution.Workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false)
-                : null;
+            async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)> GetItemsAsync()
```

Comment:
I agree.  This would also be good to go into "the list of informal IDE styles" that we were talking about @tmat .

Reply:
@CyrusNajmabadi  I would not agree to that style. It depends on the context and can't be generalized and automatically checked.

Reply:
I don't have strong preference in this particular case. I'll move the function to the end.

Reply:
> It depends on the context and can't be generalized and automatically checked.

i don't see why it can't be automatically checked.  We see if the func doesn't capture anything, or at least if everythin it captures is available at top level.  And, if so, we move it down.

Note that local functions that capture non-top-level state are also discouraged as they are often difficult to reason about and also break the flow of analyzing functions.  

IDE often takes an approach of: yes, the language allows this, but if it harms readability and maintainability, then we dissuade (though not necessarily disallow) usage of it.

Reply:
We can't automatically check it because in some cases it's more readable to have the function in the middle of the method and in other cases it's better at the end.

Reply:
> in some cases it's more readable to have the function in the middle of the method

This hasn't borne out in practice.  Usually it's a sign that the method is too complex.  And it very much impacts readability to have to know what code is going to be reentrant from later in the method and what isn't.

In these cases hte IDE approach has been to generally disallow.  Yes, that's restrictive.  However, the overall point is to have rules that *most of the time* are beneficial and can be applied consistently rather than having inconsistency that makes life more difficult.

I realize this might not be palatable.  And that you def may have coding patterns that you are more comfortable with, or which are used in other layers.  But, generally speaking, we do try to hold to this in the IDE to make sure that we have a maximally consistent style.

I'm also very aware that this is hard to learn since there is no tooling ot help out here.  I'm still very willing to help out with such a tool to try to make that happen.  It would certainly help for external contributers and new-hires as well so we don't have to manually disseminate the information on every PR :)

Reply:
Well, I guess we won't agree on this. I don't believe that all inconsistencies make life more difficult. I think rules need to be reviewed case by case. If we agree on a rule then sure, let's apply it uniformly and consistently. I don't think we need to agree on everything though and if we don't we shouldn't enforce it in any way (in PR or automatically). There has been no agreement on placement of local functions as far as I know.

Reply:
> Usually it's a sign that the method is too complex

Sometimes long does not mean complex. Breaking a long function with straightforward flow into many small functions is not always desirable.

Reply:
> I don't have strong preference in this particular case. I'll move the function to the end. 

Actually, since the function captures the locals of the containing method I'm gonna leave it as is since I want the function to only capture the parameters of the method, not the other variables whose declaration follows the function's body.